### PR TITLE
Construct a grave and bury your fallen companions or pets.

### PR DESCRIPTION
**Pull request by nexusmrsep**
```SUMMARY: Content "Construct a grave and bury your fallen companions or pets"```

#### Purpose of change
Follow-up after #28803, a highly requested feature.

#### Describe the solution
New construction type: **Dig Grave and Bury Sealed Coffin**

- this constructs a new terrain type: (fresh) grave
- requires sealed coffin, so the full burial procedure is: Construct Coffin, Seal Coffin, Dig Grave, Bury Sealed Coffin
- requires 20 rocks and two 2x4s
- whatever you put in the coffin will be buried (includes corpses and items - lost your buddy's corpse? bury his favored belongings instead)
- if you dig out a preexisting fresh grave you can exhume whatever you buried there (corpse will rot in time though), but coffin is destroyed in the process (mainly for mechanics reasons). ~~Impossible for now, because you can't dig in a place where items are located. @Night-Pryanik if its ok for mechanical reasons, can I exclude this terrain from being empty in `can_dig_here` check?~~
- this terrain can be constructed only by the player (at least for now), so exhuming will not give any debuffs or morale penalties, since I'm assuming that player is exhuming his fallen brethren for a good reason, like moving their grave somewhere safer etc. in other words: if you buried someone, you can exhume someone, no strings attached - in contrary to exploratory gravedigging on cemeteries.

#### Describe alternatives you've considered
Couple of other ways of implementation, but this was quite convenient and also fits my goal for this feature.

#### Additional context
I've reordered construction menu, since it doesn't sort by name, so options are now placed in following order for convenience:
- Build Coffin
- Seal Coffin
- Dig Grave and Bury Sealed Coffin

### CODE REVIEW

**Comment by Night-Pryanik** [comment]
> if its ok for mechanical reasons, can I exclude this terrain from being empty in can_dig_here check?

I guess there could be all sorts of weird things happening if one will try to dig a tile with items on it. Maybe add a special case for graves?

**Comment by nexusmrsep** [comment]
If by special case you mean adding graves as exception from the need of being empty then I'd be happy to do so. Or do you mean adding some extra special code handling grave contents during digging action handler?

**Comment by Rail-Runner** [comment]
Maybe give a morale bonus to spiritual characters for doing that (but only in case actual corpse is buried)?

**Comment by nexusmrsep** [comment]
@Rail-Runner good idea, I'll try to implement it tomorrow.

**Comment by Night-Pryanik** [comment]
> If by special case you mean adding graves as exception from the need of being empty then I'd be happy to do so.

This.

**Comment by codemime** [comment]
> I guess there could be all sorts of weird things happening if one will try to dig a tile with items on it. Maybe add a special case for graves?

This is a pretty recent addition (added for consistency in #29259). Before that, digging in presence of items was entirely possible. So I assume nothing terrible would happen if graves were excluded from this rule.

**Comment by codemime** [comment]
It would be nice to also be able to inscribe last words or/and an epitaph.

**Comment by nexusmrsep** [comment]
Added the exemption for digging up / exhuming a new grave. Done a live test and nothing bad happend.

> It would be nice to also be able to inscribe last words or/and an epitaph.

Great idea. Will look into it tomorrow.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/tips-tricks-and-newb-questions-reborn/16870/715

**Comment by nexusmrsep** [comment]
Added morale bonus for spiritual characters when burying a proper body, some extra messages to indicate that, and a memorial log of the event. Needs some extra testing. Run out of time to do the epitaphs but tomorrow is also a day, so I'll get back to it.

File: src/construction.cpp [hunk 1]
```diff
@@ -947,6 +950,37 @@ void construct::done_trunk_plank( const tripoint &p )
     }
 }
 
+void construct::done_grave( const tripoint &p )
+{
+    map_stack its = g->m.i_at( p );
+    for( item it : its ) {
+        if( it.is_corpse() ) {
+            if( it.get_corpse_name().empty() ) {
+                if( it.get_mtype()->has_flag( MF_HUMAN ) ) {
```
Comment (Night-Pryanik) [inline src/construction.cpp]: It seems that if it's not human corpse, then do nothing? We were supposed to bury pets too, but in this case there's no morale changes nor sad words.

File: src/construction.cpp [hunk 2]
```diff
@@ -947,6 +950,37 @@ void construct::done_trunk_plank( const tripoint &p )
...
+                    if( g->u.has_trait( trait_SPIRITUAL ) ) {
+                        g->u.add_morale( MORALE_FUNERAL, 50, 75, 1_days, 1_hours );
+                        add_msg( m_good, _( "You feel relieved after providing last rites for this human being, whose name is lost in The Cataclysm." ) );
```
Comment (Night-Pryanik) [inline src/construction.cpp]:
```suggestion
                        add_msg( m_good, _( "You feel relieved after providing last rites for this human being, whose name is lost in the Cataclysm." ) );
```
Unless I'm mistaken we don't capitalize `The`.

File: src/construction.cpp (hunk 1 above)
Comment (nexusmrsep) [reply src/construction.cpp]: Actually if your pet animal was named, it falls under further check, that does not check if corpse was a human. I double checked if pet names carry on to their corpses and if it is recognized by the bury code. It is. Will fix the descriptions of morale gain to more general though.

image

So mourn all you want, but name your pet first.
I wish I could say that no in-game dog was harmed in making this feature but... c'est la vie.

**Comment by nexusmrsep** [comment]
> It would be nice to also be able to inscribe last words or/and an epitaph.

Done. Reused graffiti code for it.

File: .gitignore [hunk 3]
```diff
@@ -170,3 +170,8 @@ Xcode/
 
 # gdb
 .gdb_history
+*.ogg
```
Comment (ZhilkinSerg) [inline .gitignore]:
```suggestion
```

File: .gitignore [hunk 4]
```diff
@@ -170,3 +170,8 @@ Xcode/
...
+*.txt
```
Comment (ZhilkinSerg) [inline .gitignore]:
```suggestion
```

File: .gitignore [hunk 5]
```diff
@@ -170,3 +170,8 @@ Xcode/
...
+data/sound/CDDA-Soundpack/musicset.json
```
Comment (ZhilkinSerg) [inline .gitignore]:
```suggestion
```

File: .gitignore [hunk 6]
```diff
@@ -170,3 +170,8 @@ Xcode/
...
+data/sound/CDDA-Soundpack/README.md
```
Comment (ZhilkinSerg) [inline .gitignore]:
```suggestion
```

File: .gitignore [hunk 7]
```diff
@@ -170,3 +170,8 @@ Xcode/
...
+data/sound/CDDA-Soundpack/soundset.json
```
Comment (ZhilkinSerg) [inline .gitignore]:
```suggestion
```

File: src/construction.cpp [hunk 8]
```diff
@@ -946,6 +950,52 @@ void construct::done_trunk_plank( const tripoint &p )
...
+                if( it.get_mtype()->has_flag( MF_HUMAN ) ) {
+                    if( g->u.has_trait( trait_SPIRITUAL ) ) {
+                        g->u.add_morale( MORALE_FUNERAL, 50, 75, 1_days, 1_hours );
+                        add_msg( m_good,
+                                 _( "You feel relieved after providing last rites for this human being, whose name is lost in the Cataclysm." ) );
+                    } else {
+                        add_msg( m_neutral, _( "You bury remains of a human, whose name is lost in the Cataclysm." ) );
+                    }
+                    g->u.add_memorial_log( pgettext( "memorial_male",
+                                                     string_format( "You buried unknown victim of the Cataclysm.", it.type_name() ) ),
```
Comment (ZhilkinSerg) [inline src/construction.cpp]:
```suggestion
                                                     string_format( "You buried unknown victim of the Cataclysm.", it.type_name() ).c_str() ),
```

File: src/construction.cpp [hunk 9]
```diff
@@ -946,6 +950,52 @@ void construct::done_trunk_plank( const tripoint &p )
...
+                                           pgettext( "memorial_female", string_format( "You buried unknown victim of The Cataclysm.",
+                                                     it.type_name() ) ) );
```
Comment (ZhilkinSerg) [inline src/construction.cpp]:
```suggestion
                                                     it.type_name() ).c_str() ) );
```

File: src/construction.cpp [hunk 10]
```diff
@@ -946,6 +950,52 @@ void construct::done_trunk_plank( const tripoint &p )
...
+                    add_msg( m_good,
+                             _( "You feel sadness, but also relief after providing last rites for %s, whose name you will keep in your memory." ),
+                             it.get_corpse_name() );
+                } else {
+                    add_msg( m_neutral,
+                             _( "You bury remains of %s, who joined uncounted masses perished in the Cataclysm." ),
+                             it.get_corpse_name() );
+                }
+                g->u.add_memorial_log( pgettext( "memorial_male", string_format( "You buried %s.",
+                                                 it.get_corpse_name() ) ),
```
Comment (ZhilkinSerg) [inline src/construction.cpp]:
```suggestion
                                                 it.get_corpse_name() ).c_str() ),
```

File: src/construction.cpp [hunk 11]
```diff
@@ -946,6 +950,52 @@ void construct::done_trunk_plank( const tripoint &p )
...
+                                       pgettext( "memorial_female", string_format( "You buried %s.", it.get_corpse_name() ) ) );
```
Comment (ZhilkinSerg) [inline src/construction.cpp]:
```suggestion
                                       pgettext( "memorial_female", string_format( "You buried %s.", it.get_corpse_name() ).c_str() ) );
```

File: data/json/construction.json [hunk 12]
```diff
@@ -1232,18 +1247,19 @@
...
+    "description": "Dig Grave and Bury Sealed Coffin",
+    "category": "DIG",
+    "required_skills": [ [ "fabrication", 0 ] ],
     "time": 60,
-    "qualities": [ [ { "id": "HAMMER", "level": 2 } ] ],
-    "components": [
-      [ [ "2x4", 12 ] ],
-      [ [ "sheet", 2 ], [ "blanket", 1 ], [ "down_blanket", 1 ], [ "fur_blanket", 1 ] ],
-      [ [ "nail", 24 ] ]
+    "qualities": [
```
Comment (ZhilkinSerg) [inline data/json/construction.json]:
```suggestion
    "qualities": [ [ { "id": "HAMMER", "level": 1 } ], [ { "id": "DIG", "level": 2 } ] ],
```

File: data/json/construction.json [hunk 13]
```diff
@@ -1232,18 +1247,19 @@
...
+      [ { "id": "HAMMER", "level": 1 } ],
```
Comment (ZhilkinSerg) [inline data/json/construction.json]:
```suggestion
```

File: data/json/construction.json [hunk 14]
```diff
@@ -1232,18 +1247,19 @@
...
+      [ { "id": "DIG", "level": 2 } ]
```
Comment (ZhilkinSerg) [inline data/json/construction.json]:
```suggestion
```

File: data/json/construction.json [hunk 15]
```diff
@@ -1232,18 +1247,19 @@
...
     ],
```
Comment (ZhilkinSerg) [inline data/json/construction.json]:
```suggestion
```

File: data/json/terrain.json [hunk 16]
```diff
@@ -10296,6 +10296,18 @@
     "color": "green",
     "move_cost": 3,
     "flags": [ "TRANSPARENT", "DIGGABLE", "MOUNTABLE", "NOCOLLIDE" ],
-    "bash": { "sound": "thump", "ter_set": "t_null", "str_min": 50, "str_max": 100, "str_min_supported": 100 }
+    "bash": { "sound": "thump", "ter_set": "t_dirt", "str_min": 50, "str_max": 100, "str_min_supported": 100 }
+  },
+   {
```
Comment (ZhilkinSerg) [inline data/json/terrain.json]:
```suggestion
  {
```

**Comment by paulenka-aleh** [comment]
I wonder what could emerge from this grave when / if #28787 is implemented :)

File: .gitignore (hunk 3 above)
Comment (nexusmrsep) [reply .gitignore]: This and more evidently got included by mistake, sorry.

File: src/game.cpp [hunk 17]
```diff
@@ -6161,7 +6161,8 @@ void game::print_graffiti_info( const tripoint &lp, const catacurses::window &w_
 
     const int max_width = getmaxx( w_look ) - column - 2;
     if( m.has_graffiti_at( lp ) ) {
-        fold_and_print( w_look, ++line, column, max_width, c_light_gray, _( "Graffiti: %s" ),
+        fold_and_print( w_look, ++line, column, max_width, c_light_gray,
+                        m.ter( lp ) == t_grave_new ? _( "Graffiti: %s" ) : _( "Incription: %s" ),
```
Comment (ZhilkinSerg) [inline src/game.cpp:6165]: This causes "Inscription" label for all graffiti in look around view if terrain is not `t_grave_new` (i.e. walls, doors, etc).

File: src/editmap.cpp [hunk 18]
```diff
@@ -683,7 +683,9 @@ void editmap::update_view( bool update_info )
         }
 
         if( g->m.has_graffiti_at( target ) ) {
-            mvwprintw( w_info, off, 1, _( "Graffiti: %s" ), g->m.graffiti_at( target ) );
+            mvwprintw( w_info, off, 1,
+                       g->m.ter( target ) == t_grave_new ? _( "Graffiti: %s" ) : _( "Incription: %s" ),
```
Comment (ZhilkinSerg) [inline src/editmap.cpp:687]: This causes "Inscription" label for all graffiti in edit map view if terrain is not `t_grave_new` (i.e. walls, doors, etc).

**Pull request merged**